src/osvr/Common/GestureComponent.cpp

```diff
@@ -0,0 +1,236 @@
+        m_gestureNameMap.registerStringID(OSVR_GESTURE_OPEN_HAND);
+        m_gestureNameMap.registerStringID(OSVR_GESTURE_CLOSED_HAND);
+    }
+
+    void GestureComponent::recordParent(Parent &dev) {
+        BOOST_ASSERT_MSG(nullptr == m_parent,
+                         "recordParent should only be called once!");
+        m_parent = &dev;
+        m_parentSet();
+
+        // add a ping handler to re-send gesture map everytime the new
+        // connection(ping) occurs
+        m_commonComponent =
+            m_getParent().addComponent(osvr::common::CommonComponent::create());
+        OSVR_TimeValue now;
+        osvrTimeValueGetNow(&now);
+        m_commonComponent->registerPingHandler([&] { m_sendGestureMap(now); });
+    }
+
+    void GestureComponent::sendGestureData(OSVR_GestureState gestureState,
+                                           std::string const &gestureName,
+                                           OSVR_ChannelCount sensor,
+                                           OSVR_TimeValue const &timestamp) {
+
+        Buffer<> buf;
+
+        StringID gestureID = m_gestureNameMap.getStringID(gestureName);
+
+        // if we just inserted new gesture ID then send gesture map
+        if (m_gestureNameMap.isUpdateAvailable()) {
+            m_sendGestureMap(timestamp);
+        }
+
+        messages::GestureRecord::MessageSerialization msg(gestureState,
+                                                          gestureID, sensor);
+        serialize(buf, msg);
+
+        m_getParent().packMessage(buf, gestureRecord.getMessageType(),
+                                  timestamp);
+    }
+
+    void GestureComponent::m_sendGestureMap(OSVR_TimeValue const &timestamp) {
+
+        Buffer<> buf;
+
+        SerializedStringMap serializedMap = m_gestureNameMap.getMap();
+
+        messages::GestureMapRecord::MessageSerialization msg(serializedMap);
+        serialize(buf, msg);
+
+        m_getParent().packMessage(buf, gestureMapRecord.getMessageType(),
+                                  timestamp);
+    }
+
+    int VRPN_CALLBACK
+    GestureComponent::m_handleGestureRecord(void *userdata,
+                                            vrpn_HANDLERPARAM p) {
+        auto self = static_cast<GestureComponent *>(userdata);
+        auto bufReader = readExternalBuffer(p.buffer, p.payload_len);
+
+        messages::GestureRecord::MessageSerialization msg;
+        deserialize(bufReader, msg);
+        auto data = msg.getData();
+        auto timestamp = util::time::fromStructTimeval(p.msg_time);
+
+        for (auto const &cb : self->m_cb) {
+            cb(data, timestamp);
+        }
+        return 0;
+    }
+
+    int VRPN_CALLBACK
+    GestureComponent::m_handleGestureMapRecord(void *userdata,
+                                               vrpn_HANDLERPARAM p) {
+        auto self = static_cast<GestureComponent *>(userdata);
+        auto bufwrap = ExternalBufferReadingWrapper<unsigned char>(
```

Comment:
Fixed. I must have missed it because I changed the first message type but not the second one.